{
 "cells": [
  {
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-11-28T11:22:31.176611Z",
     "start_time": "2024-11-28T11:22:23.415768Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import unittest\n",
    "\n",
    "# Run the Web App Routing notebook\n",
    "%run WebAppRouting.ipynb import app\n",
    "\n",
    "\n",
    "### End-to-End Testing of the Application - User logs in and manages data ###\n",
    "\n",
    "class FlaskAppTests(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        self.client = app.test_client()\n",
    "        self.client.testing = True\n",
    "        app.config['TESTING'] = True\n",
    "\n",
    "    def tearDown(self):\n",
    "        # Clean up - Delete the registered user if exists\n",
    "        self.delete_user_record('johndoe')\n",
    "        self.delete_patient_record('12345')\n",
    "        self.delete_patient_record('67890')\n",
    "        app.config['TESTING'] = False\n",
    "\n",
    "    def delete_user_record(self, username):\n",
    "        pass\n",
    "\n",
    "    def delete_patient_record(self, patient_id):\n",
    "        pass\n",
    "\n",
    "    def test_user_logs_in_and_manages_patients(self):\n",
    "        # Register a new user (if not already registered)\n",
    "        registration_data = {\n",
    "            'first_name': 'John',\n",
    "            'last_name': 'Doe',\n",
    "            'username': 'johndoe',\n",
    "            'email': 'john.doe123@example.com',\n",
    "            'password': 'SecurePass123!'\n",
    "        }\n",
    "        response = self.client.post('/register', data=registration_data, follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'SUCCESS!', response.data)\n",
    "        self.assertIn(b'YOU ARE NOW REGISTERED', response.data)\n",
    "\n",
    "        response = self.client.get('/login', follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Login', response.data)\n",
    "\n",
    "        # Login the user\n",
    "        login_data = {\n",
    "            'username': 'johndoe',\n",
    "            'password': 'SecurePass123!'\n",
    "        }\n",
    "        response = self.client.post('/login', data=login_data, follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'You are Now Logged In', response.data)\n",
    "\n",
    "        # Access view_patients page\n",
    "        response = self.client.get('/view_patients')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Stroke Dataset', response.data)\n",
    "\n",
    "        # Add a new patient\n",
    "        add_patient_data = {\n",
    "            'id' : '12345',\n",
    "            'gender': 'Male',\n",
    "            'age': 45,\n",
    "            'hypertension': 0,\n",
    "            'heart_disease': 0,\n",
    "            'ever_married': 'Yes',\n",
    "            'work_type': 'Private Sector',\n",
    "            'residence_type': 'Urban',\n",
    "            'avg_glucose_level': 85.5,\n",
    "            'bmi': 24.1,\n",
    "            'smoking_status': 'Never Smoked',\n",
    "            'stroke': 0\n",
    "        }\n",
    "        response = self.client.post('/add_patient', data=add_patient_data, follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Patient added to the database, ID number is 12345', response.data)\n",
    "\n",
    "        # Delete a patient\n",
    "        delete_patient_data = {\n",
    "        # The ID of the patient to be deleted\n",
    "            'id': '12345'\n",
    "        }\n",
    "        response = self.client.post('/delete_patient', data=delete_patient_data, follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(f'Patient with ID 12345 has been deleted'.encode(), response.data)\n",
    "\n",
    "\n",
    "\n",
    "    unittest.main(argv=[''], verbosity=2, exit=False)\n"
   ],
   "id": "initial_id",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app 'WebAppRouting'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on https://127.0.0.1:8282\n",
      "Press CTRL+C to quit\n",
      "test_user_logs_in_and_manages_patients (__main__.FlaskAppTests.test_user_logs_in_and_manages_patients) ... "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User data inserted into the database\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 9, in <module>\n",
      "    class FlaskAppTests(unittest.TestCase):\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 91, in FlaskAppTests\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 57, in test_user_logs_in_and_manages_patients\n",
      "    response = self.client.get('/view_patients')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\3241598507.py\", line 212, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 9, in <module>\n",
      "    class FlaskAppTests(unittest.TestCase):\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 91, in FlaskAppTests\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 76, in test_user_logs_in_and_manages_patients\n",
      "    response = self.client.post('/add_patient', data=add_patient_data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\3241598507.py\", line 271, in add_patient\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 9, in <module>\n",
      "    class FlaskAppTests(unittest.TestCase):\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 91, in FlaskAppTests\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 76, in test_user_logs_in_and_manages_patients\n",
      "    response = self.client.post('/add_patient', data=add_patient_data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1150, in open\n",
      "    response = self.resolve_redirect(response, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1056, in resolve_redirect\n",
      "    return self.open(builder, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\3241598507.py\", line 212, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 9, in <module>\n",
      "    class FlaskAppTests(unittest.TestCase):\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 91, in FlaskAppTests\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 85, in test_user_logs_in_and_manages_patients\n",
      "    response = self.client.post('/delete_patient', data=delete_patient_data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\3241598507.py\", line 312, in delete_patient\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 9, in <module>\n",
      "    class FlaskAppTests(unittest.TestCase):\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 91, in FlaskAppTests\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\4281634575.py\", line 85, in test_user_logs_in_and_manages_patients\n",
      "    response = self.client.post('/delete_patient', data=delete_patient_data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1150, in open\n",
      "    response = self.resolve_redirect(response, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1056, in resolve_redirect\n",
      "    return self.open(builder, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_11460\\3241598507.py\", line 212, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 1 test in 3.090s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "6eba495276ed970c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "4e006559730652c8",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
